signer/src/network/grpc_relay.rs

```diff
@@ -0,0 +1,284 @@
+//! # Relay server and client for signer communication over gRPC
+//!
+//! This module provides an implementation of the [`crate::network::MessageTransfer`]
+//! trait over gRPC.
+//!
+//! ## Features
+//!
+//! - **RelayServer**: Server implementation that listens to incoming connections and forwards incoming messages to all connected clients.
+//! - **RelayClient**: Client implementation which exposes the `MessageTransfer` interface for the signer.
+//!
+//! ## Examples
+//!
+//! - `src/bin/relay-server.rs` shows how to properly spin up a relay server.
+//!
+//! - `examples/relay-client.rs` is an an example of
+//! how to concurrently send messages to and receive messages from the server.
+
+pub mod proto {
+    tonic::include_proto!("message");
+}
+
+use std::collections::VecDeque;
+use std::net::ToSocketAddrs;
+use std::pin::Pin;
+
+use tokio::sync::broadcast;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::BroadcastStream;
+use tokio_stream::wrappers::ReceiverStream;
+use tokio_stream::Stream;
+use tokio_stream::StreamExt;
+
+use crate::codec;
+use crate::codec::Decode;
+use crate::codec::Encode;
+
+pub const CHANNEL_CAPACITY: usize = 10_000;
+
+type Msg = crate::ecdsa::Signed<crate::message::SignerMessage>;
+type MsgId = [u8; 32];
+
+/// Relay server that broadcasts incoming messages over gRPC.
+/// The internal logic is just wrapping a tokio broadcast channel.
+#[derive(Debug)]
+pub struct RelayServer {
+    pub broadcast_tx: broadcast::Sender<Result<proto::Message, tonic::Status>>,
+}
+
+impl RelayServer {
+    pub fn new() -> Self {
+        let (broadcast_tx, _) = broadcast::channel(CHANNEL_CAPACITY);
+        Self { broadcast_tx }
+    }
+
+    /// Run the server at the provided address
+    #[tracing::instrument(skip(self, addr))]
+    pub async fn serve(self, addr: impl ToSocketAddrs) -> Result<(), ConnectError> {
+        let socket_address = addr
+            .to_socket_addrs()
+            .map_err(|_| ConnectError::AddressParsing)?
+            .next()
+            .ok_or(ConnectError::AddressParsing)?;
+
+        tracing::info!(address = ?socket_address, "starting relay server");
+        tonic::transport::Server::builder()
+            .add_service(proto::relay_server::RelayServer::new(self))
+            .serve(
+                addr.to_socket_addrs()
+                    .map_err(|_| ConnectError::AddressParsing)?
+                    .next()
+                    .ok_or(ConnectError::AddressParsing)?,
+            )
+            .await?;
+
+        Ok(())
+    }
+}
+
+#[tonic::async_trait]
+impl proto::relay_server::Relay for RelayServer {
+    type BroadcastStream =
+        Pin<Box<dyn Stream<Item = Result<proto::Message, tonic::Status>> + Send>>;
+
+    #[tracing::instrument(skip(self, request))]
```

Comment:
What does this annotation do?

Reply:
It records information for subsequent logs. For example, consider this log line:

```
  2024-05-16T18:45:40.293770197Z DEBUG h2::codec::framed_write: send, frame: Data { stream_id: StreamId(1) }
    at /home/marten/.cargo/registry/src/index.crates.io-6f17d22bba15001f/h2-0.3.26/src/codec/framed_write.rs:213
    in h2::proto::connection::Connection with peer: Client
    in signer::network::grpc_relay::connect
```

While the logs stem from a third-party library, we're still getting information that this comes from the `grpc_relay::connect` function thanks to the function having this annotation.

In more technical terms, the annotation creates and enters a tracing span when the function is called. Docs on the macro (annotation): https://docs.rs/tracing/0.1.40/tracing/attr.instrument.html and on spans: https://docs.rs/tracing/latest/tracing/span/index.html